code/modules/power/debug/power_debugger.dm

```diff
@@ -0,0 +1,322 @@
+	* contained, local powernets, and general state
+	*
+	* This tool is also great for looking at local powernets and investigating where power is being consumed and the various states
+	* that `/machinery` is in. Furthermore, as a contributor, you can add additional functionality through powernet logging that's
+	* automatically built into local powernets and this tool, on production it does nothing but if you're trying to investigate something
+	* specific, you can track whatever you want on a local net and view it through this UI module.
+
+*/
+/datum/ui_module/powernet_debugger
+	name = "Powernet Debugger"
+	/// A hard reference to the currnet regional powernet we are analyzing in the UI screen
+	var/datum/regional_powernet/selected_regional_net = null
+	/// A hard reference to the current local powernet we are analyzing in the UI screen
+	var/datum/local_powernet/selected_local_net = null
+
+	/// a list of UIDs pointing to powernets we have selected, all powernets in this list will show up as tabs in the UI module, great for bouncing between multiple things
+	var/list/selected_powernets = list()
+	/// The current screen we're on, just used to distinguish looking at the list of regional powernets vs. the detailed view of the net we have selected
+	var/debug_screen = PW_DEBUG_SCREEN_POWERNETS
+	/// A cache of machines/power-machines that we're listing. We don't want to send a billion icons to the UI-module, so we only send one of each by caching stuff
+	var/list/power_icon_cache = list()
+
+	// # Filter Tools
+	/// When TRUE, removes pipes from the local powernet machine list
+	var/filter_pipes = FALSE
+	/// When TRUE, removes walls lights from the local powernet machine list
+	var/filter_lights = FALSE
+	/// When TRUE, removes APCs from the regional powernet power machine list
+	var/filter_apcs = FALSE
+	/// When TRUE, removes terminals from the regional powernet power machine list
+	var/filter_terminals = FALSE
+	/// When TRUE, removes off-station powernets from the initial regional powernet list
+	var/filter_non_station_powernets = TRUE // enabled by default b/c most contributors and admins wont care about off-station nets
+
+/datum/ui_module/library_manager/ui_state(mob/user)
+	return GLOB.admin_state
+
+/datum/ui_module/powernet_debugger/ui_interact(mob/user, datum/tgui/ui = null)
+	ui = SStgui.try_update_ui(user, src, ui)
+	if(!ui)
+		ui = new(user, src, "PowernetDebugger", name)
+		ui.autoupdate = TRUE
+		ui.open()
+
+/// Returns a list of data on regional powernets in the world to be used in ui_act that
+/datum/ui_module/powernet_debugger/proc/get_all_powernets()
+	var/list/powernets = list()
+	for(var/datum/regional_powernet/powernet in SSmachines.powernets)
+		if(length(powernet.cables) < PW_MIN_CABLE_COUNT)
+			continue // we don't give a shit about tiny powernets
+		var/obj/structure/cable/test_cable = powernet.cables[1]
```

Comment:
are we guaranteed to have at least one cable in the powernet at this point or should there be a check here for that?

Reply:
yes. Powernets nuke themselves if they're devoid of cables. There has to be a cable in order for a powernet datum to exist because propagation of the powernet itself starts from a cable.